beacon-chain/state/stategen/replayer.go

```diff
@@ -0,0 +1,347 @@
+package stategen
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/pkg/errors"
+	types "github.com/prysmaticlabs/eth2-types"
+	"github.com/prysmaticlabs/prysm/beacon-chain/core/helpers"
+	"github.com/prysmaticlabs/prysm/beacon-chain/db"
+	"github.com/prysmaticlabs/prysm/beacon-chain/state"
+	"github.com/prysmaticlabs/prysm/encoding/bytesutil"
+	"github.com/prysmaticlabs/prysm/proto/prysm/v1alpha1/block"
+	"github.com/sirupsen/logrus"
+	"go.opencensus.io/trace"
+)
+
+var ErrFutureSlotRequested = errors.New("cannot replay to future slots")
+var ErrNoCanonicalBlockForSlot = errors.New("none of the blocks found in the db slot index are canonical")
+var ErrNoBlocksBelowSlot = errors.New("no blocks found in db below slot")
+var ErrInvalidDBBlock = errors.New("invalid block found in database")
+var ErrGenesisBlockNotFound = errors.New("canonical genesis block not found in db")
+
+// HistoryAccessor describes the minimum set of database methods needed to support the ReplayerBuilder
+type HistoryAccessor interface {
+	HighestSlotBlocksBelow(ctx context.Context, slot types.Slot) ([]block.SignedBeaconBlock, error)
+	GenesisBlock(ctx context.Context) (block.SignedBeaconBlock, error)
+	Block(ctx context.Context, blockRoot [32]byte) (block.SignedBeaconBlock, error)
+	StateOrError(ctx context.Context, blockRoot [32]byte) (state.BeaconState, error)
+}
+
+// CanonicalChecker is used by the Replayer to find the highest canonical block in the database <= the target slot
+// in practice this should be satisfied by a type that uses the fork choice store.
+type CanonicalChecker interface {
+	IsCanonical(ctx context.Context, blockRoot [32]byte) (bool, error)
+}
+
+// CurrentSlotter is used to ensure requestedSlot < currentSlot, to fail fast on requests to replay to future slots
+type CurrentSlotter interface {
+	CurrentSlot() types.Slot
+}
```

Comment:
It's hard for me to understand the purpose of this interface from the doc. How does it fail fast on requests when it `is used to ensure requestedSlot < currentSlot`? Doesn't this mean that the interface will make sure we are on the happy path, and thus *not* failing?

Reply:
by "fail fast" I mean that the Replayer will check to see if the requested slot is in the past, and if so return an error, before doing anything else like querying the database.

Reply:
OK gotcha. But since this interface is exported, shouldn't the doc explain what the interface is supposed to do (get the current slot) instead of how it's supposed to be used? Other packages might have a different use case than replaying she state.

Reply:
fair enough, "how it's used internally" is TMI, I can talk about "what it does" instead. however, this interface is intended to define what some code in this package *expects* vs what it *returns*. So I wouldn't necessarily expect it to be used in other ways.